app/boot_control/grub.py

```diff
@@ -120,49 +118,74 @@ class GrubHelper:
 
     @classmethod
     def update_entry_rootfs(
-        cls, grub_cfg: str, *, kernel_ver: str, rootfs_str: str
+        cls,
+        grub_cfg: str,
+        *,
+        kernel_ver: str,
+        rootfs_str: str,
+        start: int = 0,
     ) -> Optional[str]:
-        """Read in grub_cfg and return updated one.
-        NOTE: we only update normal entry for the <kernel_ver>,
-        recovery mode or other special entry will be ignored.
+        """Read in grub_cfg, update all entries' rootfs with <rootfs_str>,
+            and then return the updated one.
 
         Params:
             grub_cfg: input grub_cfg str
             kernel_ver: kernel version str for the target entry
-            rootfs_str: a str that indicates which rootfs device to use
+            rootfs_str: a str that indicates which rootfs device to use,
+                like root=UUID=<uuid>
         """
         new_entry_block: Optional[str] = None
         entry_l, entry_r = None, None
+        is_updated = False
 
         # loop over normal entry, find the target entry,
         # and then replace the rootfs string
-        for entry in cls.menuentry_pa.finditer(grub_cfg):
+        for entry in cls.menuentry_pa.finditer(grub_cfg, start):
             entry_l, entry_r = entry.span()
             entry_block = entry.group()
-            # NOTE: ignore recovery entry
-            if entry.group("kernel_ver") == kernel_ver and not entry.group("other"):
-                if linux_line := cls.linux_pa.search(entry_block):
-                    linux_line_l, linux_line_r = linux_line.span()
-                    new_linux_line = "%s%s%s" % (
-                        linux_line.group("left"),
-                        rootfs_str,
-                        linux_line.group("right"),
-                    )
-                    new_entry_block = (
-                        f"{entry_block[:linux_line_l]}"
-                        f"{new_linux_line}"
-                        f"{entry_block[linux_line_r:]}"
+            # parse the entry block
+            if _linux := cls.linux_pa.search(entry_block):
+                if _linux.group("ver") == kernel_ver:
+                    linux_line_l, linux_line_r = _linux.span()
+                    _new_linux_line, _count = cls.rootfs_pa.subn(
+                        rootfs_str, _linux.group()
                     )
-                    break
+                    if _count == 1:
+                        # replace rootfs string
+                        new_entry_block = "%s%s%s" % (
+                            entry_block[:linux_line_l],
+                            _new_linux_line,
+                            entry_block[linux_line_r:],
+                        )
+                        is_updated = True
+                        break
+
+        if is_updated and new_entry_block is not None:
```

Comment:
is_update flag is required? I think only evaluating new_entry_blck is enough.

Reply:
Yes you are right, the `is_updated` flag is the leftover for the previous non-recursive implementation, which is not required anymore.

Reply:
resolved in [e2c12af](https://github.com/tier4/ota-client/pull/152/commits/e2c12af04d2de903852be87cfe06d81560b23bef).